www/utils/screenshot.ts

```
@@ -1,32 +1,63 @@
+    throw new Error("Invalid URL");
+  }
+  return parsedUrl;
 }
 
-const browser = await launch();
-const page = await browser.newPage(url);
-await page.waitForNetworkIdle();
-const raw = await page.screenshot();
+export function generateFilePaths(
+  id: string,
+): { image2x: string; image1x: string } {
+  return {
+    image2x: `./www/static/showcase/${id}2x.jpg`,
+    image1x: `./www/static/showcase/${id}1x.jpg`,
+  };
+}
+
+export async function captureScreenshot(
+  url: string,
+  id: string,
+): Promise<void> {
+  const browser = await launch();
+  try {
+    const page = await browser.newPage(url);
+    await page.waitForNetworkIdle();
+    const raw = await page.screenshot();
 
-await browser.close();
+    const image2x = await Image.decode(raw);
+    const { image2x: path2x, image1x: path1x } = generateFilePaths(id);
 
-// convert to jpeg
-const image2x = await Image.decode(raw);
-await Deno.writeFile(
-  `./www/static/showcase/${id}2x.jpg`,
-  await image2x.encodeJPEG(80),
-);
+    await Deno.writeFile(path2x, await image2x.encodeJPEG(80));
+    const image1x = image2x.resize(image2x.width / 2, Image.RESIZE_AUTO);
+    await Deno.writeFile(path1x, await image1x.encodeJPEG(80));
+  } finally {
+    await browser.close();
+  }
+}
 
-const image1x = image2x.resize(image2x.width / 2, Image.RESIZE_AUTO);
-await Deno.writeFile(
-  `./www/static/showcase/${id}1x.jpg`,
-  await image1x.encodeJPEG(80),
-);
+// only run the script if it's the main module
+if (import.meta.main) {
+  try {
+    const [url, id] = validateArgs(Deno.args);
+    validateUrl(url);
+    await captureScreenshot(url, id);
+    // deno-lint-ignore no-console
+    console.log(`Screenshot saved as ${id}1x.jpg and ${id}2x.jpg`);
+  } catch (error) {
+    const message = error instanceof Error ? error.message : String(error);
+    // deno-lint-ignore no-console
+    console.log(`Error: ${message}`);
+    Deno.exit(1);
+  }
```

This `catch` block seems redundant. Why not just let the program throw naturally?

---

+1 on erroring naturally.